File: frontend/src/types/course.ts
export type CourseStatus = 'open' | 'draft' | 'closed'

export interface ClassInfo {
  id: number
  className: string
  schedule: string
  location: string
  studentCount: number
}

/**
 * 课程视图模型。
 * 列表接口（GET /courses）返回除 classes 外的全部字段，班级数量用 classCount；
 * 详情接口（GET /courses/:id）额外返回 classes 明细。
 */
export interface Course {
  id: number
  code: string
  name: string
  credit: number
  hours: number
  semester: string
  departmentId: number
  department: string
  teacherId: number
  teacherName: string
  description: string
  /** 培养目标 / 适用专业：Sprint 2 课程表扩展列，当前后端不返回，渲染时按空值降级 */
  objective?: string
  major?: string
  classes?: ClassInfo[]
  classCount: number
  studentCount: number
  resourceCount: number
  status: CourseStatus
}

export interface CourseQuery {
  semester?: string
  departmentId?: number
  teacherId?: number
  status?: CourseStatus | ''
  keyword?: string
  page?: number
  pageSize?: number
}

/** 与后端 CourseUpsertReq 逐字对齐的提交结构 */
export interface CourseUpsertPayload {
  code: string
  name: string
  departmentId: number
  teacherId: number
  semester: string
  credit: number
  hours: number
  description: string
  status: CourseStatus
}

/** 表单模型：下拉未选择时保持空串，提交前收敛为 CourseUpsertPayload */
export interface CourseFormModel {
  id?: number
  code: string
  name: string
  credit: number
  hours: number
  semester: string
  departmentId: number | ''
  teacherId: number | ''
  description: string
  status: CourseStatus
}

File: backend/internal/dto/course.go
package dto

// CourseListQuery 是 GET /courses 的查询参数。
type CourseListQuery struct {
	Semester     string `form:"semester"`
	DepartmentID uint64 `form:"departmentId"`
	TeacherID    uint64 `form:"teacherId"`
	Status       string `form:"status" binding:"omitempty,oneof=open draft closed"`
	Keyword      string `form:"keyword"`
	Page         int    `form:"page"`
	PageSize     int    `form:"pageSize"`
}

// CourseListItem 是课程列表行。
// 说明：credit/hours/description/teacherId/departmentId 为满足前端列表与编辑回填的字段，
// 仍全部来自 courses 主表，不引入冗余汇总列。
type CourseListItem struct {
	ID            uint64 `json:"id"`
	Code          string `json:"code"`
	Name          string `json:"name"`
	Credit        int    `json:"credit"`
	Hours         int    `json:"hours"`
	Description   string `json:"description"`
	TeacherID     uint64 `json:"teacherId"`
	TeacherName   string `json:"teacherName"`
	DepartmentID  uint64 `json:"departmentId"`
	Department    string `json:"department"`
	Semester      string `json:"semester"`
	ClassCount    int    `json:"classCount"`
	StudentCount  int    `json:"studentCount"`
	ResourceCount int    `json:"resourceCount"`
	Status        string `json:"status"`
}

// ClassInfoDTO 是课程详情中的开课班级。
type ClassInfoDTO struct {
	ID           uint64 `json:"id"`
	ClassName    string `json:"className"`
	Schedule     string `json:"schedule"`
	Location     string `json:"location"`
	StudentCount int    `json:"studentCount"`
}

// CourseDetail 是 GET /courses/:id 的响应。
type CourseDetail struct {
	CourseListItem
	Classes []ClassInfoDTO `json:"classes"`
}

// CourseUpsertReq 是 POST/PUT /courses 的请求体。
type CourseUpsertReq struct {
	Code         string `json:"code" binding:"required,max=12,coursecode"`
	Name         string `json:"name" binding:"required,max=128"`
	DepartmentID uint64 `json:"departmentId" binding:"required"`
	TeacherID    uint64 `json:"teacherId" binding:"required"`
	Semester     string `json:"semester" binding:"required,max=16"`
	Credit       int    `json:"credit" binding:"required,min=1,max=6"`
	Hours        int    `json:"hours" binding:"required,min=16,max=128"`
	Description  string `json:"description" binding:"max=500"`
	Status       string `json:"status" binding:"required,oneof=open draft closed"`
}
